server/src/main/java/org/elasticsearch/health/node/ShardLimitsHealthIndicatorService.java

```diff
@@ -0,0 +1,210 @@
+        return createIndicator(
+            finalStatus,
+            symptomBuilder.toString(),
+            buildDetails(normalNodes.result, frozenNodes.result),
+            indicatorImpacts,
+            diagnoses
+        );
+    }
+
+    private StatusResult calculateForNormalNodes(HealthMetadata.ShardLimits shardLimits) {
+        int maxShardsPerNodeSetting = shardLimits.maxShardsPerNode();
+        var result = ShardLimitValidator.checkShardLimitForNormalNodes(maxShardsPerNodeSetting, 5, 1, clusterService.state());
+
+        if (result.canAddShards() == false) {
+            return new StatusResult(HealthStatus.RED, result);
+        }
+
+        result = ShardLimitValidator.checkShardLimitForNormalNodes(maxShardsPerNodeSetting, 10, 1, clusterService.state());
+        if (result.canAddShards() == false) {
+            return new StatusResult(HealthStatus.YELLOW, result);
+        }
+
+        return new StatusResult(HealthStatus.GREEN, result);
+    }
+
+    private StatusResult calculateForFrozenNodes(HealthMetadata.ShardLimits shardLimits) {
+        int maxShardsPerNodeSetting = shardLimits.maxShardsPerNodeFrozen();
+        var result = ShardLimitValidator.checkShardLimitForFrozenNodes(maxShardsPerNodeSetting, 5, 1, clusterService.state());
+        if (result.canAddShards() == false) {
+            return new StatusResult(HealthStatus.RED, result);
+        }
+
+        result = ShardLimitValidator.checkShardLimitForFrozenNodes(maxShardsPerNodeSetting, 10, 1, clusterService.state());
+        if (result.canAddShards() == false) {
+            return new StatusResult(HealthStatus.YELLOW, result);
+        }
+
+        return new StatusResult(HealthStatus.GREEN, result);
+    }
+
+    static HealthIndicatorDetails buildDetails(ShardLimitValidator.Result normalNodes, ShardLimitValidator.Result frozenNodes) {
+        return (builder, params) -> {
+            builder.startObject();
+            {
+                builder.startObject(NORMAL_GROUP + "_nodes");
+                builder.field("max_shards_in_cluster", normalNodes.maxShardsInCluster());
+                if (normalNodes.currentUsedShards().isPresent()) {
+                    builder.field("current_used_shards", normalNodes.currentUsedShards().get());
+                }
+                builder.endObject();
+            }
+            {
+                builder.startObject(FROZEN_GROUP + "_nodes");
+                builder.field("max_shards_in_cluster", frozenNodes.maxShardsInCluster());
+                if (frozenNodes.currentUsedShards().isPresent()) {
+                    builder.field("current_used_shards", frozenNodes.currentUsedShards().get());
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+            return builder;
+        };
+    }
+
+    private HealthIndicatorResult unknownIndicator() {
+        return createIndicator(
+            HealthStatus.UNKNOWN,
+            "No number of shards limits data.",
```

Comment:
If this will be seen by a user (will Cloud UI expose this?) I think we can word it better as I had to read it a few times. 

"No data regarding shard limits was available"? "Unable to determine shard limit status"?